tree-sitter/src/CodeGen/Deserialize.hs

```diff
@@ -55,17 +55,18 @@ instance FromJSON MkDatatype where
       Just subtypes   -> pure (SumType type' named subtypes)
 
 -- | Transforms list of key-value pairs to a Parser
-parseKVPairs :: NonEmpty (Text, Value) -> Parser (NonEmpty (String, MkField))
+parseKVPairs :: NonEmpty (Text, Value) -> Parser (NonEmpty MkField)
 parseKVPairs = traverse go
-  where go :: (Text, Value) -> Parser (String, MkField)
+  where go :: (Text, Value) -> Parser MkField
         go (t,v) = do
           v' <- parseJSON v
-          pure (unpack t, v')
+          pure $ v' { fieldName = Just (unpack t) }
 
 data MkField = MkField
   { fieldRequired :: MkRequired
-  , fieldTypes     :: [MkType]
+  , fieldTypes    :: [MkType]
   , fieldMultiple :: MkMultiple
+  , fieldName     :: Maybe String
```

Comment:
We make this a `Maybe String` so that the JSON blobs (and the FromJSON instance) don’t necessarily have to provide a fieldName value upon first parse. In practice, `parseKVPairs` will ensure that this is always `Just`.

Reply:
This is the basic reason we went with the pairs in the first place. I’d personally much rather have the datatypes reflect the possible values accurately and live with tuples than move the name back into `Field` and introduce margin for error.

Up to @aymannadeem ultimately, but I’m personally 👎

Reply:
On the other hand, we could also just make `Field`’s `FromJSON` instance handle the pair itself. Wouldn’t be able to enjoy the generic implementation any more, but we’d get the precise typing + idiomatic usage. Win-win?

Reply:
@robrix The `FromJSON` instance handling the pair is exactly what I’ve done. The `Maybe String` ensures that we load from the JSON correctly (since the fieldName key won’t be there), but after that loading happens (line 63) we update the fieldName with that yielded from the pair.

I feel pretty strongly that avoiding tuples in record types is the right thing to do. It’s visually ugly and encourages proliferation of function parameters/otiose uses of `curry` and `uncurry`. If both elements of the tuple are required to complete a given computation, they should be packaged together, rather than passed around ad-hoc.

Reply:
@patrickt: The JSON value we would get is the one representing the pair, no? We only get the value representing the second field of the pair because that’s what the current instance decides to do.